### PR TITLE
Rwblickhan/#215 add rocket token command

### LINKED ISSUE #215: Add `/rocket token` command
**Please give a one-sentence summary of the feature you would like to see.**
Add a `/rocket token` command that generates and returns a signed [JSON Web Token](https://jwt.io/) for the user, which can then be used to authenticate with a potential HTTP API in the future.

### PR DESCRIPTION
**Pull request by rwblickhan**
# Pull Request

## Description

Give a brief description of your changes:
Add a `/rocket token` command that generates and presents a signed JWT for the user, for use with a potential future HTTP API.

## Testing

If testing this change requires extra setup, please document it here:
See picture of output in comments below.

## Ticket(s)

Closes #215

### CODE REVIEW

**Comment by rwblickhan** [comment]
screen shot 2019-02-10 at 8 49 46 pm

^ Output.

File: command/commands/token.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        self.expiry = config.get_expiry()
+        self.signing_key = config.get_signing_key()
+
+    def get_name(self):
+        """Return the command's name."""
+        return self.command_name
+
+    def get_desc(self):
+        """Return a description of the command."""
+        return self.desc
```
Comment (bobheadxi) [inline command/commands/token.py]: this is more of an aside, but TIL lots of getters is a thing in python, I thought the meme was that java's getters are half of its verbosity?

File: command/commands/token.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+    def handle(self, _command, user_id):
+        """Handle request for token."""
+        logging.debug("Handling token command")
```
Comment (bobheadxi) [inline command/commands/token.py]: one thing I've discovered is that logs like this are *super* unhelpful, because you end up getting a mumbo jumbo of logs with no way of associating series of events. For example, 2 users calling the same endpoint might generate logs like this:

```
DEBUG handling
DEBUG handling
INFO handling with option
INFO handling without option
INFO something
WARNING oh no
INFO success
INFO success
```

gets weird trying to tell who's who - it might be helpful to have some sort of request ID that propagates down to functions and can be included in logs as a field

File: command/commands/token.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+            'iss': 'ubclaunchpad:rocket2',
+            'iat': datetime.utcnow(),
+            'user_id': user_id,
+            'permissions': user.get_permissions_level().value
+        }
+        token = jwt.encode(payload, self.signing_key, algorithm='HS256') \
+            .decode('utf-8')
+        return "This is your token:\n```\n{}".format(token) + \
+               "\n```\nKeep it secret! Keep it safe!\n" \
+               "It will expire at {}.".format(expiry), \
```
Comment (bobheadxi) [inline command/commands/token.py]: this feels a little tricky to read 😢 

it might help to have a set of slack/markdown util functions like this:

```python
def code(str):
    return "`{}`".format(str)

def code_block(str):
    return "```\n{}\n```".format(str)

def quote(str):
    return "> {}\n".format(str)

def emph(str):
    return "*{}*".format(str)
```

etc, then this could be:

```python
return "This is your token: {}".format(slack_fmt.code_block(token))
```

Comment (bobheadxi) [reply command/commands/token.py]: (since I imagine such formatting will probably be used frequently throughout rocket)

File: Pipfile [hunk 4]
```diff
@@ -14,6 +14,8 @@ pygithub = "*"
 toml = "*"
 structlog = "*"
 colorama = "*"
+pyjwt = "*"
+flask-limiter = "*"
```
Comment (bobheadxi) [inline Pipfile]: this doesn't look like it's used in this PR yet?

Comment (rwblickhan) [reply Pipfile]: No, I was planning to start working on that but figured I should get this reviewed first. I don't think it's worth removing at this point though, because it definitely _will_ be used at some point.

File: command/commands/token.py (hunk 1 above)
Comment (rwblickhan) [reply command/commands/token.py]: I'm not sure that a lot of getters is actually a thing in Python - thoughts, @cheukyin699?

File: command/commands/token.py (hunk 3 above)
Comment (rwblickhan) [reply command/commands/token.py]: Definitely a good idea, but low priority I would say - I'll make a ticket.

Comment (rwblickhan) [reply command/commands/token.py]: (Unless _you_ want to make a ticket. 😛)

Comment (bobheadxi) [reply command/commands/token.py]: haha sure ill do it, but yeah i actually feel this is a pretty high priority, since the longer you wait the more you'll have to replace and the less likely anyone will want to do it 😋

Comment (bobheadxi) [reply command/commands/token.py]: see #218

Comment (rwblickhan) [reply command/commands/token.py]: Doesn't matter if they want to do it if it's their assigned task 😉 

And while it is somewhat hard to read, I don't think it's _that_ important to fix - I think it's relatively obvious what the output is supposed to be, even if the exact formatting is a bit hard to parse.

Comment (rwblickhan) [reply command/commands/token.py]: Thanks for making the new ticket though!

Comment (bobheadxi) [reply command/commands/token.py]: Hard to parse == easy to make mistakes 😰 I updated the example:

```py
return "This is your token: {}".format(slackfmt.code_block(token)) + \
    "Keep it secret! Keep it safe!\n" + \
    "It will expire at {}.".format(expiry), \
    200
```

I'd say removing the "noise" of various markdown annotations makes it a lot clearer

File: command/commands/token.py (hunk 2 above)
Comment (rwblickhan) [reply command/commands/token.py]: Honestly `user_id` would probably work for that - that's passed directly from Slack to us, so it's as good as a request ID (and I highly doubt somebody would try spamming commands to such a point we can't tell their requests apart).

File: command/commands/token.py (hunk 3 above)
Comment (rwblickhan) [reply command/commands/token.py]: But you've already done the work of figuring out what it does, and we never need to touch it again, right 😛

Comment (bobheadxi) [reply command/commands/token.py]: Or:

```py
return """
This is your token: {}".format(slackfmt.code_block(token))
Keep it secret! Keep it safe!
It will expire at {}.".format(expiry)
    """,
    200
```

(untested, but i think that would work)

Comment (bobheadxi) [reply command/commands/token.py]: 😢

File: command/commands/token.py (hunk 2 above)
Comment (bobheadxi) [reply command/commands/token.py]: that is, until we have users programmatically accessing APIs with tokens and whatnot

Comment (rwblickhan) [reply command/commands/token.py]: But this is _strictly_ for the Slack command - somebody accessing programmatically won't ever arrive here.

Comment (bobheadxi) [reply command/commands/token.py]: oh, i was under the assumption that whatever API you guys build my hook into the command framework to avoid rewriting functionality, like:

```sh
rocket2.ubclaunchpad.com/user/edit?member=rwblickhan
# /rocket user edit --member=rwblickhan
```

Comment (rwblickhan) [reply command/commands/token.py]: Probably something like that - though I was specifically referring to `/rocket token`, which should reaaaally only be accessible via Slack 😛

Comment (bobheadxi) [reply command/commands/token.py]: tricky though - ideally an API would return proper JSON structures (we return plain text in inertia and https://github.com/ubclaunchpad/inertia/pull/546 was... not fun to write...)

I might consider having these functions return objects with relevant information, like

```
{ token: "", expiry: "" }
```

then write a thin layer above the commands that handles things in one of two ways:

* throw the object straight to the caller if via API
* format object into a pretty object if via Slack

Comment (bobheadxi) [reply command/commands/token.py]: > specifically referring to `/rocket token`, which should reaaaally only be accessible via Slack

oh yeah that's fair, I guess this is going on a bit more of a tangent about things in general 😋 just throwing some thoughts out here

Comment (rwblickhan) [reply command/commands/token.py]: > then write a thin layer above the commands that handles things in one of two ways:
> 
>     * throw the object straight to the caller if via API
> 
>     * format object into a pretty object if via Slack

I was thinking something along these lines, yeah. That would require a bit of refactoring, but hopefully not _too_ much.

File: command/commands/token.py (hunk 1 above)
Comment (cheukyin699) [reply command/commands/token.py]: I think he's referring to the `@property` decorator, where you could make a function with that decorator and have slightly easier access to the variables within. Also see `@var.setter` and `@var.getter`.

A slight caveat is the classes that implement this must be children of the mighty `object` class. But nothing a bit of refactoring can't do.

Comment (rwblickhan) [reply command/commands/token.py]: Sounds like another ticket. 😛

File: command/commands/token.py (hunk 3 above)
Comment (michaelenglo) [reply command/commands/token.py]: Fyi, I did some stack overflow on this topic. We can write it like this:
```
'''line {0}
line {1}
line {2}'''.format(1,2,3)
```
and this will print `'line 1\nline 2\nline 3'`.

Might help make it less ugly(?) 😕

Comment (cheukyin699) [reply command/commands/token.py]: If we use that, we would destroy the indentation. In my opinion, it would make it uglier.

Comment (rwblickhan) [reply command/commands/token.py]: I tend to agree... it would work, but @bobheadxi's utility solution is probably "cleaner."

Actually, now that I think about it, there's probably going to be other Launch Pad Python projects that would benefit from those functions. Perhaps we should break it out into a separate repo...?

Comment (bobheadxi) [reply command/commands/token.py]: you mean a package like `slackfmt`? 😍

Comment (rwblickhan) [reply command/commands/token.py]: Slack and possibly general Markdown formatting as well, yeah. That should be relatively easy to build/test.

Comment (cheukyin699) [reply command/commands/token.py]: > *Grumble grumble* you teams and your multiple repositories/projects

> Something something something lisp

It sounds like a small side-project more though. Might work if we add more features.

Comment (bobheadxi) [reply command/commands/token.py]: `mdfmt` 🤔

Comment (rwblickhan) [reply command/commands/token.py]: Maybe `mdformatter` instead? 😛

Comment (bobheadxi) [reply command/commands/token.py]: grumble grumble library names should short for easier imports grumble grumble

Comment (rwblickhan) [reply command/commands/token.py]: Aren't you using an IDE with autocomplete? 😛 

Besides, `fmt` makes me think of auto-formatters like `gofmt` and `rustfmt`, which this very much is not.

Comment (bobheadxi) [reply command/commands/token.py]: yeah but go's format-printing standard library is called `fmt`

```go
import "fmt"

func main() {
    var truth = "mdfmt is best"
    fmt.Printf("%s", truth)
}
```

Comment (rwblickhan) [reply command/commands/token.py]: Hmm... I imagine we'd put both Markdown and Slack formatting into the same library though... so perhaps `lpfmt`? 😛

Comment (bobheadxi) [reply command/commands/token.py]: Slack formatting is only a subset of Markdown, right? theres nothing in slack formatting that's not in MD

the only thing I'm aware of is italics are `_blah_` instead of `*blah*`, and bold is `*blah*` instead of `**blah**`

Comment (rwblickhan) [reply command/commands/token.py]: 😱 Seems like a pretty big difference!

Comment (bobheadxi) [reply command/commands/token.py]: potentially cool: a flag that toggles between MD and Slack mode, so if you try to `as_table()` with the Slack mode on, isntead of a markdown table you'll get a markdown-table-in-code-block

Comment (rwblickhan) [reply command/commands/token.py]: See, I was thinking it would be easier to just have separate modules (in the same package) for Markdown and Slack formatting...

Comment (bobheadxi) [reply command/commands/token.py]: I wrote a small slack module a long time ago - the entirety of its formatting can be encompassed in <50 lines though:

https://gist.github.com/bobheadxi/92737e761d60072ca905dfa3dd096cc3

Comment (rwblickhan) [reply command/commands/token.py]: Yeah, this doesn’t have to be complicated, I just don’t think we should have it stuck in the Rocket 2 repository rather than, say, on PyPI 😛

Comment (bobheadxi) [reply command/commands/token.py]: oh i was under the impression you were thinking of spinning it off into a new repo 😋 well in that case keep it simple 😅

Comment (rwblickhan) [reply command/commands/token.py]: Yeah, I would spin it off into a new repo, just to keep it separate/distinct from all the Rocket 2 stuff.

Comment (rwblickhan) [reply command/commands/token.py]: I don’t think it has to be particularly complicated, even if it’s in a new repo.

File: command/core.py [hunk 5]
```diff
@@ -9,13 +10,14 @@
...
 
-    def __init__(self, db_facade, bot, gh_interface):
+    def __init__(self, db_facade, bot, gh_interface, token_config):
         """Initialize the dictionary of command handlers."""
         self.__commands = {}
         self.__facade = db_facade
         self.__bot = bot
         self.__github = gh_interface
         self.__commands["user"] = UserCommand(self.__facade, self.__github)
+        self.__commands["token"] = TokenCommand(self.__facade, token_config)
```
Comment (jacqueschen1) [inline command/core.py]: Is it time to finally start looking into meta-classes? :p

Comment (rwblickhan) [reply command/core.py]: You don't have a task anymore, right? You could definitely start looking into that now, if you'd like.

Comment (jacqueschen1) [reply command/core.py]: I thought I was supposed to do team commands as well? But yea

File: tests/command/commands/token_test.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                             "secret",
+                             algorithms='HS256')
+        assert decoded['user_id'] == user.get_slack_id()
+        assert decoded['permissions'] == Permissions.team_lead.value
+        assert ret_code == 200
+
+    def test_handle_admin(self):
+        """Test handle() when given a user with admin permissions."""
+        user = User("U12345")
+        user.set_permissions_level(Permissions.admin)
```
Comment (cheukyin699) [inline tests/command/commands/token_test.py]: Didn't you make an admin user in `setUp()`?

Comment (rwblickhan) [reply tests/command/commands/token_test.py]: Oops, my mistake, it should be done here. I'll fix that.

File: command/commands/token.py [hunk 7]
```diff
@@ -0,0 +1,77 @@
+"""Command to obtain signed authentication token."""
+import jwt
+import logging
+
+from datetime import datetime
+from model.permissions import Permissions
+
+
+class TokenCommand:
+    """Token command model class."""
```
Comment (ravenCondol) [inline command/commands/token.py]: I noticed that we don't initialize parsers here like in the other commands which confuses me ? Or is this something that is to be added later

Comment (rwblickhan) [reply command/commands/token.py]: We don't need parsers, there's no subcommands/options.

**Comment by cheukyin699** [comment]
oh that's what that button does

File: command/commands/token.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+    desc = "Generate a signed token for use with the HTTP API"
+    permission_error = "You do not have the sufficient " \
+                       "permission level for this command!"
+    lookup_error = "Requesting user not found!"
+
+    def __init__(self, db_facade, config):
+        """
+        Initialize TokenCommand.
+
+        :param db_facade: Database connection
```
Comment (cheukyin699) [inline command/commands/token.py]: Add config object to parameters

File: command/commands/token.py (hunk 3 above)
Comment (cheukyin699) [inline command/commands/token.py]: This success message should be a static constant:

```python
success_msg = "This is your token..."
# ...
success_msg.format(token, expiry, ...)
```

File: factory/__init__.py [hunk 9]
```diff
@@ -17,17 +22,27 @@ def make_core(config, gh=None):
...
             config['slack']['creds_path'])['api_token']
         github_api_token = toml.load(
             config['github']['creds_path'])['api_token']
         github_organization = config['github']['organization']
         gh = GithubInterface(Github(github_api_token), github_organization)
+        if os.path.isfile(config['auth']['signing_key_path']):
+            signing_key = open(config['auth']['signing_key_path']).read()
+        else:
+            signing_key = ''.join(
+                random.choice(string.ascii_lowercase) for _ in range(24))
```
Comment (cheukyin699) [inline factory/__init__.py]: Abstract into another function?

File: command/commands/token.py (hunk 8 above)
Comment (rwblickhan) [reply command/commands/token.py]: Whoops, will fix.

File: command/commands/token.py (hunk 3 above)
Comment (rwblickhan) [reply command/commands/token.py]: Will fix.

File: factory/__init__.py (hunk 9 above)
Comment (rwblickhan) [reply factory/__init__.py]: Sure.

**Comment by rwblickhan** [comment]
Issue for documentation in #244.
